src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/KeyValuePairConverter.cs

```diff
@@ -3,31 +3,27 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Generic;
-using System.Text.Encodings.Web;
+using System.Diagnostics;
 
 namespace System.Text.Json.Serialization.Converters
 {
-    internal sealed class KeyValuePairConverter<TKey, TValue> : JsonValueConverter<KeyValuePair<TKey, TValue>>
+    internal sealed class KeyValuePairConverter<TKey, TValue> :
+        SmallObjectWithParameterizedConstructorConverter<KeyValuePair<TKey, TValue>, TKey, TValue, object, object>
```

Comment:
I'm not following. If we want to invoke the ctor on KeyValuePair then we can just call it directly -- why the need for the indirection?

Reply:
The problem is we have to duplicate all the logic for async (de)serialization (from the parameterized ctor converter) and make sure they stay in sync. The write side is relatively straightforward, but reading is more cumbersome - managing `state.Current.ObjectState`. `state.Current.PropertyState`, remembering previously parsed key/value ctor args (with `ArgumentState`) when we need to fetch more data etc.

If we had parameterized ctor support in v1 and didn't enforce that the payload have precisely 2 properties, one for the key and one for the value, we wouldn't have needed the KeyValuePair converter at all. The KeyValuePair has to exist only to do the property validation.

Reply:
I see. We want to support Streams properly (meaning not using read-ahead) so we are leveraging the newly added ctor support to do that.